File: main/framework/controllers/workflows.py
"""Workflow HTTP routes — thin handlers that delegate to WorkflowQueryService.

Each endpoint is a thin shell:
  1. Validate the request via Pydantic schemas.
  2. Call one (or a few) ``WorkflowQueryService`` methods.
  3. Let the global RFC 7807 handlers in ``main.py`` map service-layer
     ``NotFoundError`` → 404 and ``ServiceError`` → 500 / 400.

Business logic lives in:
  - ``WorkflowQueryService``     — CRUD, stats, trigger (this controller)
  - ``WorkflowEngine``           — async DAG execution (spawned by trigger)

The router defines 7 routes:
  POST   /                       (create,         201)
  GET    /                       (list,           200)
  GET    /stats                  (get_stats,      200)
  GET    /{workflow_id}          (get,            200)
  PUT    /{workflow_id}          (update,         200)
  DELETE /{workflow_id}          (delete,         204)
  POST   /{workflow_id}/trigger  (trigger,        202)

The re-export shim at ``api/workflows.py`` re-publishes this ``router`` under
the original import path so ``main.py`` and any other consumer keep working
unchanged.

DI strategy
-----------
Per Wave 4.3 the controllers use ``Depends(get_service(...))`` exclusively.
The DB session is sourced from the injected ``WorkflowRepository`` via its
``_session()`` context manager — the same pattern ``SessionService`` and the
new ``conversations.py`` controller use for per-call db access.
"""

from __future__ import annotations

import asyncio
import contextlib
import logging

from fastapi import APIRouter, Depends, status
from main.framework.core.infrastructure.container import get_container, get_service
from main.framework.repositories.workflow_repo import WorkflowRepository
from main.framework.services.workflow_query_service import WorkflowQueryService
from pydantic import BaseModel

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/workflows", tags=["workflows"])


# ---------------------------------------------------------------------------
# Request / response schemas
# ---------------------------------------------------------------------------


class WorkflowCreate(BaseModel):
    name: str
    description: str | None = None
    nodes: list[dict] = []
    edges: list[dict] = []
    trigger_type: str | None = "manual"
    config: dict | None = {}


class WorkflowUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    nodes: list[dict] | None = None
    edges: list[dict] | None = None
    trigger_type: str | None = None
    config: dict | None = None


class WorkflowTrigger(BaseModel):
    params: dict | None = {}


# ---------------------------------------------------------------------------
# Async trigger runner — used by the trigger endpoint
# ---------------------------------------------------------------------------


async def _run_workflow_async(
    workflow_id: str,
    params: dict,
    execution_id: str,
    container,
) -> None:
    """Background task that drives workflow execution.

    Mirrors the legacy ``api/triggers.py::_run_workflow_async`` but pulls
    everything from the DI container (so it participates in test
    dependency-overrides and future container-managed resources).
    """
    wf_repo = container.workflow_repo
    exec_repo = container.execution_repo
    try:
        workflow = wf_repo.get(workflow_id)
        if not workflow:
            return

        with contextlib.suppress(Exception):
            exec_repo.update_execution(execution_id, status="running")

        from main.framework.core.workflow.node_executors.agent_executor import _resolve_agent_name
        for node in workflow.nodes or []:
            agent = _resolve_agent_name(node)
            with contextlib.suppress(Exception):
                exec_repo.create_node(
                    execution_id=execution_id,
                    node_id=node["id"],
                    agent=agent,
                    status="pending",
                    input=params,
                )

        engine = container.create_workflow_engine(workflow_id, params, execution_id=execution_id)
        await engine.execute()
    except Exception as e:  # noqa: BLE001 — last-resort guard for the background task
        logger.error("Workflow execution failed: %s", e, exc_info=True)
        with contextlib.suppress(Exception):
            exec_repo.update_execution(execution_id, status="failed")


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_workflow(
    payload: WorkflowCreate,
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """Create a new workflow. Returns the created workflow (full detail)."""
    with wf_repo._session() as db:
        return service.create_workflow(payload.model_dump(), db)


@router.get("")
async def list_workflows(
    skip: int = 0,
    limit: int = 1000,
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """List workflows (summary view, newest first)."""
    with wf_repo._session() as db:
        return service.list_workflows(db, skip=skip, limit=limit)


@router.get("/stats")
async def get_workflow_stats(
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """Get aggregated workflow execution statistics."""
    with wf_repo._session() as db:
        return service.get_workflow_stats(db)


@router.get("/{workflow_id}")
async def get_workflow(
    workflow_id: str,
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """Get a workflow by id. 404 via the global NotFoundError handler."""
    with wf_repo._session() as db:
        return service.get_workflow(workflow_id, db)


@router.put("/{workflow_id}")
async def update_workflow(
    workflow_id: str,
    payload: WorkflowUpdate,
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """Update a workflow. Re-validates the DAG if nodes/edges change."""
    with wf_repo._session() as db:
        return service.update_workflow(workflow_id, payload.model_dump(exclude_none=True), db)


@router.delete("/{workflow_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workflow(
    workflow_id: str,
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """Delete a workflow and cascade-delete its executions."""
    with wf_repo._session() as db:
        service.delete_workflow(workflow_id, db)
    return None


@router.post("/{workflow_id}/trigger", status_code=status.HTTP_202_ACCEPTED)
async def trigger_workflow(
    workflow_id: str,
    payload: WorkflowTrigger,
    wf_repo: WorkflowRepository = Depends(get_service(WorkflowRepository)),
    service: WorkflowQueryService = Depends(get_service(WorkflowQueryService)),
):
    """Trigger a workflow execution asynchronously.

    The service creates the pending ``WorkflowExecution`` row and returns the
    id; this handler then schedules the async runner via the container's
    ``create_workflow_engine`` factory so the work runs out-of-band.
    """
    params = payload.params or {}
    with wf_repo._session() as db:
        execution_id = service.trigger_workflow(workflow_id, params, db)
    container = get_container()
    asyncio.create_task(_run_workflow_async(workflow_id, params, execution_id, container))
    return {"execution_id": execution_id}

File: src/webui/src/api/workflows.ts
/**
 * Typed wrappers for the workflow CRUD HTTP routes.
 *
 * Mount points (see `src/main/api/v1/workflows.py`):
 *
 * GET    /api/v1/workflows               → listWorkflows
 * POST   /api/v1/workflows               → createWorkflow
 * GET    /api/v1/workflows/{id}          → getWorkflow
 * PUT    /api/v1/workflows/{id}          → updateWorkflow
 * DELETE /api/v1/workflows/{id}          → deleteWorkflow
 * POST   /api/v1/workflows/{id}/trigger  → triggerWorkflow
 */

import { ROUTES } from './contract'
import { apiGet, apiPost, apiPut, apiDelete } from './http'
import type {
  Workflow,
  WorkflowMeta,
} from '../domain/workflow'

/* ─── Payload types ───────────────────────────────────────────────── */

export interface CreateWorkflowPayload {
  name: string
  description?: string
  nodes?: Workflow['nodes']
  edges?: Workflow['edges']
  trigger_type?: Workflow['trigger_type']
  config?: Workflow['config']
}

export interface UpdateWorkflowPayload {
  name?: string
  description?: string
  nodes?: Workflow['nodes']
  edges?: Workflow['edges']
  trigger_type?: Workflow['trigger_type']
  config?: Workflow['config']
}

export interface TriggerParams {
  [key: string]: unknown
}

export interface TriggerResult {
  execution_id: string
}

/* ─── Workflow CRUD ───────────────────────────────────────────────── */

/** List workflows (summary view, newest first). `GET /api/v1/workflows?offset=&limit=`. */
export function listWorkflows(
  offset: number = 0,
  limit: number = 1000,
  signal?: AbortSignal,
): Promise<WorkflowMeta[]> {
  const qs = new URLSearchParams({ offset: String(offset), limit: String(limit) })
  return apiGet<WorkflowMeta[]>(`${ROUTES.workflows.list}?${qs.toString()}`, signal)
}

/** Get a single workflow by id (full detail). `GET /api/v1/workflows/{id}`. Throws on 404. */
export function getWorkflow(id: string, signal?: AbortSignal): Promise<Workflow> {
  return apiGet<Workflow>(ROUTES.workflows.get(id), signal)
}

/** Create a new workflow. Backend validates the DAG (rejects cycles, > 50 nodes).
 * `POST /api/v1/workflows` — 201 Created. */
export function createWorkflow(data: CreateWorkflowPayload, signal?: AbortSignal): Promise<Workflow> {
  return apiPost<Workflow>(ROUTES.workflows.create, data, signal)
}

/** Update a workflow; undefined fields are omitted (backend `exclude_none=True`).
 * Re-validates the DAG when `nodes` or `edges` change. `PUT /api/v1/workflows/{id}`. */
export function updateWorkflow(
  id: string,
  data: UpdateWorkflowPayload,
  signal?: AbortSignal,
): Promise<Workflow> {
  return apiPut<Workflow>(ROUTES.workflows.update(id), data, signal)
}

/** Delete a workflow and cascade-delete its executions.
 * `DELETE /api/v1/workflows/{id}` — 204 No Content. */
export function deleteWorkflow(id: string, signal?: AbortSignal): Promise<void> {
  return apiDelete<void>(ROUTES.workflows.delete(id), signal)
}

/** Trigger a workflow run asynchronously. `POST /api/v1/workflows/{id}/trigger` — 202.
 * @param params Free-form input forwarded to the engine (defaults to `{}`). */
export function triggerWorkflow(
  id: string,
  params?: TriggerParams,
  signal?: AbortSignal,
): Promise<TriggerResult> {
  return apiPost<TriggerResult>(
    ROUTES.workflows.trigger(id),
    { params: params ?? {} },
    signal,
  )
}
